tests/cpp/test_multidevice_matmul.cpp

```diff
@@ -0,0 +1,302 @@
+#include <ops/all_ops.h>
+#include <preseg_passes/allocation_order_inference.h>
+#include <preseg_passes/optimization_pass.h>
+#include <scheduler/mma_utils.h>
+#include <scheduler/utils.h>
+#include <tests/cpp/multidevice.h>
+#include <tests/cpp/validator.h>
+#include <transform_replay.h>
+#include <transform_rfactor.h>
+
+namespace nvfuser {
+
+class DistributedMatmulTest : public MultiDeviceTest {
+ protected:
+  DistributedMatmulTest() : optimization_guard_(false) {
+    DisableOptionsGuard::getCurOptions().set(DisableOption::MatmulExprEval);
+  }
+
+  void SetUp() {
+    if (!deviceMajorMinorCheck(8)) {
+      GTEST_SKIP() << "Distributed matmul tests require Ampere or newer";
+    }
+    MultiDeviceTest::SetUp();
+    num_devices = communicator->size();
+  }
+
+  MultiDeviceExecutorParams executor_params{
+      .use_fusion_executor_cache = true,
+      .skip_auto_scheduling = false,
+      .cache_fusion_executor = false};
+  int num_devices;
+
+  std::tuple<at::Tensor, at::Tensor, at::Tensor> getAtenInputOutputs(
+      MmaLayout layout,
+      int M,
+      int N,
+      int K) {
+    at::manual_seed(0);
+    int local_rank = communicator->local_rank();
+    c10::ScalarType type = c10::ScalarType::Half;
+    auto a = matmulAtInput2D(
+        layout, TensorMatmulPos::A, type, M, N, K, 0, local_rank);
+    auto b = matmulAtInput2D(
+        layout, TensorMatmulPos::B, type, M, N, K, 0, local_rank);
+    auto c =
+        atMatmul(a.to(at::kDouble), b.to(at::kDouble), layout).to(at::kFloat);
+    return std::make_tuple(a, b, c);
+  }
+
+ private:
+  preseg_passes::OptimizationPassGuard<preseg_passes::AllocationDomainPass>
+      optimization_guard_;
+  DisableOptionsGuard option_guard_;
+};
+
+TEST_F(DistributedMatmulTest, LayoutTN_NoComms) {
+  // MmaLayout::TN A(T), B(N), C(T)
+  // A and C are sharded on dimension M
+  // Tests local matmul with no communication
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  auto mesh = DeviceMesh::createForNumDevices(communicator->size());
+
+  int M = 256, N = 64, K = 64;
+  int Mo = num_devices;
+  int Mi = M / Mo;
+  std::vector<int> a_shape = {Mo, Mi, K};
+  std::vector<int> b_shape = {N, K};
+
+  TensorView* a = makeContigTensor(3, DataType::Half); // (Mo,Mi,K)
+  TensorView* b = makeContigTensor(2, DataType::Half); // (N,K)
+  TensorView* a_b = broadcast(a, {false, false, true, false}); // (Mo,Mi,b,K)
+  TensorView* b_b = broadcast(b, {true, true, false, false}); // (b,b,N,K)
+  TensorView* ab = mul(a_b, b_b); // (Mo,Mi,N,K)
+  TensorView* c = sum(ab, {-1}); // (Mo,Mi,N,r)
+
+  fusion->addInput(a);
+  fusion->addInput(b);
+  fusion->addOutput(c);
+
+  // Sharding M dimension
+  auto all_sharded_tvs = {a, a_b, b_b, ab, c};
+  for (auto tv : all_sharded_tvs) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+    tv->setDeviceMesh(mesh);
+  }
+  b->setDeviceMesh(mesh);
+
+  auto [a_, b_, c_] = getAtenInputOutputs(MmaLayout::TN, M, N, K);
```

Comment:
Just a note: the codebase has been using `name_` for class fields, so this had confused me until I realized the underscore is to indicate the tensor corresponding to TensorView `name`. This is OK unless you figure out a better way of naming.